tap/api/tcp_stream.go

```diff
@@ -1,43 +1,45 @@
-package tap
+package api
 
 import (
 	"encoding/binary"
 	"sync"
+	"time"
 
 	"github.com/google/gopacket"
 	"github.com/google/gopacket/layers" // pulls in all layers decoders
 	"github.com/google/gopacket/reassembly"
-	"github.com/up9inc/mizu/tap/api"
-	"github.com/up9inc/mizu/tap/diagnose"
+	"github.com/up9inc/mizu/tap/api/diagnose"
 )
 
 /* It's a connection (bidirectional)
  * Implements gopacket.reassembly.Stream interface (Accept, ReassembledSG, ReassemblyComplete)
  * ReassembledSG gets called when new reassembled data is ready (i.e. bytes in order, no duplicates, complete)
- * In our implementation, we pass information from ReassembledSG to the tcpReader through a shared channel.
+ * In our implementation, we pass information from ReassembledSG to the TcpReader through a shared channel.
  */
-type tcpStream struct {
-	id              int64
+type TcpStream struct {
```

Comment:
I see we moved TcpStream and TcpReader to the api package, which is a good initiative. However, currently those structs coupled with the traditional tapper and gopacket. There is no clean way to use them from other tappers, like TLS and future tappers.

For example, there is no `TcpState *reassembly.TCPSimpleFSM` in TLS, and faking it is ugly. Indeed many members of those structs are not relevant for TLS.

Maybe it need to be an interface, with some public functions, allowing every implementation to implement its specific.

Reply:
The struct is moved to `tap/api` package without changing its fields so much. `tap/api` package is the package shared between `tap` and the protocol dissectors. I couldn't understand the example of `TcpState        *reassembly.TCPSimpleFSM` field. That field is not populated in here and it does not need to be populated:

https://github.com/up9inc/mizu/blob/0c5678976c71ae46996f18aef30ec0631d96f71d/tap/tlstapper/tls_poller.go#L171-L174

I agree that `TcpStream` struct has fields which are specific to `gopacket`. I think there can be a better abstraction based on commonalities between the network sniffer and tracers. But this PR is not aiming to achieve that abstraction. It rather moves the code from here to there.

Reply:
Yes, I understand the PR doesn't introduced those structs. Yet moving a struct to an "api" package, make it an api.

Reply:
Is the problem exported fields? If so, we need define some functions to make them private again.

Reply:
If the motivation of the PR is to allow the extensions to call to `CloseOtherProtocolDissectors`, we should define an interface with this method, instead of exposing a lot of internal logic.

In tls, we have tls_reader.go, which is the equivalent to tcp_reader.go. After this change, tls suddenly need to have both, without a good reason. 

---

I still think that we should unify the tcp_reader and tls_reader, and move the core logic to the api. But this should be done in a generic way, not just by copy paste a single implementation to the api.

Reply:
As long as there is this logic:

https://github.com/up9inc/mizu/blob/0c5678976c71ae46996f18aef30ec0631d96f71d/tap/passive_tapper.go#L72-L75

The eBPF - OpenSSL tracer has nothing to do with the protocol dissector API. Because it's not complying with the dissection of multiple application layer protocols criteria. I would first remove that logic then try to unify everything under `tap/api`. But that's off-topic for this PR.

If you think the exported structs `TcpReaderDataMsg`, `TcpReader` and `TcpStream` and their exported fields are too much. For the sake of good code, we can make them private and export only the methods.

> There's no performance difference between exported and unexported fields/methods/etc.

As I stated above, the TLS tapper is neither compliant with multiple protocol support nor it's compliant with the `TcpStream` struct. So, I don't see a clear interface that unifies everything at this moment which can easily be implemented. Even if there is such an interface possible, it's outside of the scope of this PR.

Reply:
The tls use the http extension via the extension api, It doesn't matter if it use a single extension or many of them.
Those 3 struct relevant to pcap capture only - TcpReaderDataMsg, TcpReader and TcpStream.

Unification, while possible, takes time.

 ---

The whole motivation of the PR is to be able to call `CloseOtherProtocolDissectors` from the extensions, which is possible using an interface. I can't understand why we need to expose and break the encapsulation

Reply:
This commit converted pcap specific structs to pcap specific interfaces, its better to stay with the structs.

A single interface, like this once, may do the work without exposing pcap logic to the api.
```
type ExtensionHandler interface {
	Close(Protocol protocol)
}
```

The Dissect method may get it as another parameter, keeping the code the exact way it was, while achieving the same goal.

Reply:
or maybe keep the interfaces in `tap/api` but move the structs to `tap`?

Reply:
Did that in here https://github.com/up9inc/mizu/commit/7543db557c6a76835632fc4266146d5fda7193d5 but ended up with an import cycle:

```
building mizu agent ..
package command-line-arguments
	imports github.com/up9inc/mizu/tap
	imports github.com/up9inc/mizu/tap/tlstapper
	imports github.com/up9inc/mizu/tap: import cycle not allowed
make: *** [Makefile:36: agent] Error 1
```

so I had to move them to a new package called `tap/tcp`.

`CloseOtherProtocolDissectors` needs to be a pointer receiver which holds references to all the client and server `TcpReader`s such that it can close all of the other `TcpReaderDataMsg` channels. That indicates `TcpStream` is the most suitable struct to have that method. That's why `TcpStream` is passed to `Dissect` and its method signature is changed.

Reply:
That required the protocol dissectors being depend on `tap` package here https://github.com/up9inc/mizu/commit/367a812ec64359db76bf1180c3dd20003f411f76 So, I guess I will have to revert https://github.com/up9inc/mizu/commit/7543db557c6a76835632fc4266146d5fda7193d5 and https://github.com/up9inc/mizu/commit/367a812ec64359db76bf1180c3dd20003f411f76

Reply:
I'll try to elaborate.

In `tap/api/api.go` add 
```
type ExtensionHandler interface {
	Close(protocol Protocol)
}
```
Changed `Dissector.Dissect` signature to looks like 
```
Dissect(b *bufio.Reader, progress *ReadProgress, capture Capture, isClient bool, tcpID *TcpID, counterPair *CounterPair, superTimer *SuperTimer, superIdentifier *SuperIdentifier, emitter Emitter, options *TrafficFilteringOptions, reqResMatcher RequestResponseMatcher, handler ExtensionHandler) error
```

In `tap/tcp_reader.go` introduce a new struct 
```
type tcpExtensionHandler struct {
	reader *tcpReader
}

func (t *tcpExtensionHandler) Close(protocol api.Protocol) {
	fmt.Printf("Close the extension using both protocl + reader information")
}
```

Then change the call dissect like this:
```
h.extension.Dissector.Dissect(b, h.progress, h.parent.origin, h.isClient, h.tcpID, h.counterPair, h.superTimer,
		h.parent.superIdentifier, h.emitter, filteringOptions, h.reqResMatcher, &tcpExtensionHandler{reader: h})
```

---

Its just an example, names, locations etc may be anywhere. But the idea is to be able to let the extension communicate its status, without the need to give it all the internal information.

Reply:
@dvdlevanon but close happens on arrays of TCP readers:

```
	clients         []api.TcpReader
	servers         []api.TcpReader
```

It's not just single reader:


```go
func (t *tcpStream) CloseOtherProtocolDissectors(protocol *api.Protocol) {
	t.Lock()
	defer t.Unlock()

	if t.protoIdentifier.IsClosedOthers {
		return
	}

	t.protoIdentifier.Protocol = protocol

	for i := range t.clients {
		reader := t.clients[i]
		if reader.GetExtension().Protocol != t.protoIdentifier.Protocol {
			reader.Close()
		}
	}
	for i := range t.servers {
		reader := t.servers[i]
		if reader.GetExtension().Protocol != t.protoIdentifier.Protocol {
			reader.Close()
		}
	}

	t.protoIdentifier.IsClosedOthers = true
}
```

Reply:
If that's the case, we can give `tcpExtensionHandler` the tcpStream instead of tcpReader. We can also initialize it in the tcpStream instead of tcpReader.

Reply:
@dvdlevanon OK then what's the difference between passing `tcpStream` inside `tcpExtensionHandler` argument or passing `tcpStream` argument?

Reply:
We don't pass `tcpStream` insdie `tcpExtensionHandler`.
We pass `ExtensionHandler interface`, which is an interface containing a single method.

For the API this interface should be opaque, they don't know and shouldn't count on its concrete implementation.

Reply:
@dvdlevanon there were 11 arguments in the `Dissect` method signature. With that it becomes 12. Most of the arguments are actually either sourced from `tcpStream` or `tcpReader` that's why I decided to pass a `tcpReader` instance.

I don't think with;

```
h.extension.Dissector.Dissect(b, h.progress, h.parent.origin, h.isClient, h.tcpID, h.counterPair, h.superTimer,
		h.parent.superIdentifier, h.emitter, filteringOptions, h.reqResMatcher, &tcpExtensionHandler{reader: h})
```

It will be cleaner.

Reply:
I mean if we're going to pass the reader with `&tcpExtensionHandler{reader: h})` anyway and the reader holds the information of all the other arguments then why do we just don't pass the reader like it's already implemented in this PR?

Reply:
Yes I agree we should do a refactor and create a struct for those arguments.
This struct should be generic and shouldn't contain any pcap/tls related members.

If we have the time to do this refactor, great. If not, a 12th param is better than breaking the encapsulation.

Reply:
@dvdlevanon if you think that;

https://github.com/up9inc/mizu/blob/9aed5b294129e69f4fcce5bcfd9f0bbc3f90f433/tap/tlstapper/tls_reader.go#L10-L16

should implement;

https://github.com/up9inc/mizu/blob/9aed5b294129e69f4fcce5bcfd9f0bbc3f90f433/tap/api/api.go#L415-L432

I can do it as well. If I do that, is it OK?

Reply:
What `GetParent() TcpStream` returns for TlsReader?

Reply:
@dvdlevanon returns `TcpStream`

https://github.com/up9inc/mizu/blob/9aed5b294129e69f4fcce5bcfd9f0bbc3f90f433/tap/api/api.go#L434-L450

Reply:
This breaks the encapsulation, again.

tcpStream contains information not relevant to extensions, and moreover, if they'll accidentally use it it may break tls.

The fact it called `dummy` by itself, proves the point.

Reply:
> tcpStream contains information not relevant to extensions, and moreover, if they'll accidentally use it it may break tls.

It's relevant to the protocol dissectors because they elevate into TCP stream level and close the other readers with `CloseOtherProtocolDissectors` call.

https://github.com/up9inc/mizu/blob/9aed5b294129e69f4fcce5bcfd9f0bbc3f90f433/tap/api/api.go#L435-L437

are required by [this interface](https://github.com/google/gopacket/blob/3eaba08943250fd212520e5cff00ed808b8fc60a/reassembly/tcpassembly.go#L362-L380) defined in `gopacket`.

> The fact it called `dummy` by itself, proves the point.

It's suffixed `Dummy` because the `tlstapper` package does not make any use of it other than setting the `origin` field.

Reply:
Upon looking at again, I think this;

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/tlstapper/tls_poller.go#L25-L34

corresponds to;

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/api/api.go#L436-L452

since it holds references to `api.TcpReader`;

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/tlstapper/tls_poller.go#L27

just like;

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/tcp/tcp_stream.go#L37-L38

also;

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/tlstapper/tls_poller.go#L29

vs.

https://github.com/up9inc/mizu/blob/2f1e98d7fe226ed0243c72b09713b7ead70f5a01/tap/tcp/tcp_stream.go#L41

So, I can make `tlsPoller` implement `TcpStream`? WDYT?

Reply:
`It's relevant to the protocol dissectors because they elevate into TCP stream level and close the other readers with CloseOtherProtocolDissectors call.`

With the interface they achieve the same goal, without actually closing them by themselves. Thats why interfaces invented.

---

The fact that `tcpStream` needs to implement `gopacket` internals, is yet another prove why it shouldn't be part of the api.

---

`It's suffixed Dummy because the tlstapper package does not make any use of it other than setting the origin field`

Yet is another prove why it shouldn't be part of the api

Reply:
Adding yet another argument and defining a specific, smaller interface achieves the same goal. I didn't say it does not achieve the same goal. But TCP is a common property for the network sniffer, eBPF tracer and protocol dissectors. So, the TCP related definitions can reside in `tap/api` IMHO.

`TcpStream` interface defines methods that're related to TCP reassembly which is again a common property for the network sniffer and eBPF tracer (because they both should build the TCP stream). I guess `gopacket`'s TCP reassembly is not used in `tlstapper` package because the data/packets does not come in out of order. So, those methods not being required inside `tlstapper` does not make them completely irrelevant.

Reply:
OK that's make sense. 
Let's go with it.

Reply:
With https://github.com/up9inc/mizu/commit/97072427f67b8c00982a1abc4c99060eab8971e4 `tap/api` does not depend on `gopacket` and `gopacket` specific methods are now removed from `TcpStream` interface. So, it looks like this now;

https://github.com/up9inc/mizu/blob/97072427f67b8c00982a1abc4c99060eab8971e4/tap/api/api.go#L433-L449

Reply:
@dvdlevanon with https://github.com/up9inc/mizu/commit/b0fbaec2f4b534421775975326ff50615df2ab48 I have made `tlsPoller` implement `TcpStream` interface and remove the call to `NewTcpStreamDummy` method.